zebrad/src/components/inbound.rs

```diff
@@ -0,0 +1,156 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use futures::{future::FutureExt, stream::TryStreamExt};
+use tokio::sync::oneshot;
+use tower::{buffer::Buffer, util::BoxService, Service, ServiceExt};
+
+use zebra_network as zn;
+use zebra_network::AddressBook;
+use zebra_state as zs;
+
+type Outbound = Buffer<BoxService<zn::Request, zn::Response, zn::BoxError>, zn::Request>;
+type State = Buffer<BoxService<zs::Request, zs::Response, zs::BoxError>, zs::Request>;
+
+pub type SetupData = (Outbound, Arc<Mutex<AddressBook>>);
+
+pub struct Inbound {
+    // invariant: outbound, address_book are Some if network_setup is None
+    //
+    // why not use an enum for the inbound state? because it would mean
+    // match-wrapping the body of Service::call rather than just expect()ing
+    // some Options.
```

Comment:
We could add our own `expect` methods to the enum if we wanted to avoid this invariant

alternatively, we could implement deref to the address book in a way that panics if its the wrong enum variant :)

Reply:
:+1: for now, I think just using options is fine.